### PR TITLE
BUG: Index.str.partition not nan-safe (#23558)

### LINKED ISSUE #23558: BUG: Index.str.partition not nan-safe
While working on #23167, I found a corner case where `Index.str.partition` and `Index.str.rpartition` break in the presence of NaNs. I do not believe this is intentional (and it's not mentioned in the docs):

```
>>> import pandas as pd
>>> pd.Index(['a', 'b', 'c']).str.partition(' ')  # works
MultiIndex(levels=[['a', 'b', 'c'], [''], ['']],
           labels=[[0, 1, 2], [0, 0, 0], [0, 0, 0]])
>>>
>>> pd.Index(['a', np.nan, 'c']).str.partition(' ')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\ProgramData\Miniconda3\envs\pandas-dev\lib\site-packages\pandas\core\strings.py", line 2391, in partition
    return self._wrap_result(result, expand=expand)
  File "C:\ProgramData\Miniconda3\envs\pandas-dev\lib\site-packages\pandas\core\strings.py", line 2014, in _wrap_result
    out = MultiIndex.from_tuples(result, names=name)
  File "C:\ProgramData\Miniconda3\envs\pandas-dev\lib\site-packages\pandas\core\indexes\multi.py", line 1326, in from_tuples
    arrays = list(lib.to_object_array_tuples(tuples).T)
  File "pandas/_libs/src\inference.pyx", line 1559, in pandas._libs.lib.to_object_array_tuples
TypeError: object of type 'float' has no len()
```

### LINKED ISSUE #23677: API/BUG: Index.str.split(expand=True) not nan-safe
This is similar to #23558 and shares the same underlying reason: #23578
Found through extensive testing introduced in #23582 (which itself is a split off from #23167)

```
>>> values = ['a', np.nan, 'c']
>>> pd.Series(values).str.split(' ')
0    [a]
1    NaN
2    [c]
dtype: object
>>> pd.Series(values).str.split(' ', expand=True)
     0
0    a
1  NaN
2    c
>>> pd.Index(values).str.split(' ')
Index([['a'], nan, ['c']], dtype='object')
>>> pd.Index(values).str.split(' ', expand=True)
Traceback (most recent call last):
[...]
TypeError: object of type 'float' has no len()
```

### PR DESCRIPTION
**Pull request by meiermark**
- [x] closes #23558 and 
closes #23677
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -2274,7 +2274,7 @@ def to_object_array_tuples(rows: list):
 
     k = 0
     for i in range(n):
-        tmp = len(rows[i])
+        tmp = 1 if is_float(rows[i]) else len(rows[i])
```
Comment (gfyoung) [inline pandas/_libs/lib.pyx]: I presume you're meaning to check if we have `np.nan` ?  If so, let's be more explicit (e.g. use `pandas.isnull`).  You could also use `is_list_like` from `pandas/core/dtypes/inference.py`.

Comment (meiermark) [reply pandas/_libs/lib.pyx]: Yes, indeed. That's the case I wanted to handle. I fixed it in the next commit.

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a whatsnew note (bug fixes in string section). lgtm. otherwise

**Comment by jreback** [comment]
cc @h-vetinari if you want to have a look.

**Comment by h-vetinari** [comment]
@jreback @meiermark 

Not sure if this is closed intentionally or permanently. It's a valid proposal to change the output of `to_object_array_tuples`, but primarily, the problem should be tackled through the MultiIndex constructor, which is the actual problem for `.str.partition`, see #23578. I've got a solution already as part of #23167, but this is currently blocked by / waiting for #23582.

**Comment by meiermark** [comment]
Sorry for the confusion. I destroyed my branch and needed to restore it. Unfortunately, the pull request was closed by this.

File: doc/source/whatsnew/v0.24.0.txt [hunk 2]
```diff
@@ -1192,7 +1192,7 @@ Numeric
 Strings
 ^^^^^^^
 
--
+- BUG Index.str.partition not nan-safe (:issue:`23558`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: can you use ``:func:`Index.str.partition` ``, and format like the other issues

File: pandas/_libs/lib.pyx [hunk 3]
```diff
@@ -2274,7 +2274,7 @@ def to_object_array_tuples(rows: list):
 
     k = 0
     for i in range(n):
-        tmp = len(rows[i])
+        tmp = 1 if util.is_nan(rows[i]) else len(rows[i])
```
Comment (jreback) [inline pandas/_libs/lib.pyx]: actually, use checknull here, ``is_nan`` is not strict enough as it doesn't check ``None`` (same below as well)

File: pandas/tests/test_strings.py [hunk 4]
```diff
@@ -2524,28 +2524,30 @@ def test_partition_series(self):
...
-        exp = Index(np.array([('a_b', '_', 'c'), ('c_d', '_', 'e'), (
-            'f_g', '_', 'h')]))
+        exp = Index(np.array([('a_b', '_', 'c'), ('c_d', '_', 'e'),
+                              ('f_g', '_', 'h'), np.nan]))
         tm.assert_index_equal(result, exp)
         assert result.nlevels == 1
 
         result = values.str.partition('_')
-        exp = Index([('a', '_', 'b_c'), ('c', '_', 'd_e'), ('f', '_', 'g_h')])
+        exp = Index([('a', '_', 'b_c'), ('c', '_', 'd_e'),
```
Comment (jreback) [inline pandas/tests/test_strings.py]: use a ``None`` here for one of the ``np.nan``

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.0.txt (hunk 2 above)
Comment (meiermark) [reply doc/source/whatsnew/v0.24.0.txt]: I tried to follow the style of the other messages in the next commit. I hope it's correct now.

File: pandas/_libs/lib.pyx (hunk 3 above)
Comment (meiermark) [reply pandas/_libs/lib.pyx]: you are right

File: pandas/tests/test_strings.py (hunk 4 above)
Comment (meiermark) [reply pandas/tests/test_strings.py]: added multiple None's in the tests

**Comment by h-vetinari** [comment]
@meiermark 
This should very likely also solve #23677. Could you have a look, and maybe adapt the tests accordingly as well?

**Comment by meiermark** [comment]
@h-vetinari I extended some tests to check #23677. Your assumption was right, it's also resolved.
@jreback can you have a look again? The current state should include the required changes.

File: doc/source/whatsnew/v0.24.0.rst [hunk 5]
```diff
@@ -1261,7 +1261,7 @@ Numeric
 Strings
 ^^^^^^^
 
--
+- BUG in :meth:`Index.str.partition` was not nan-safe (:issue:`23558`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.rst]: can you following existing style BUG -> Bug

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
does this close both issues at the top of the PR? can you annotate a test where they are covered. also add to the whatsnew both issue numbers.

**Comment by meiermark** [comment]
@jreback Yes, this PR closes both issues. Added comments to the tests (links to the corresponding issues) and a new whatsnew message.

File: doc/source/whatsnew/v0.24.0.rst (hunk 5 above)
Comment (meiermark) [reply doc/source/whatsnew/v0.24.0.rst]: followed the existing style, now
Whoops. Good catch, thanks. Fixed in 25d5743

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @meiermark keep em coming!

**Pull request merged**